### ISSUE TITLE
[eslint-plugin-lit-a11y]: implement lang

### ISSUE THREAD

**Issue by bennypowers** [open-wc/open-wc#1925]
see: https://raw.githubusercontent.com/jsx-eslint/eslint-plugin-jsx-a11y/master/docs/rules/lang.md
although this rule may not be relevant to most lit-html use cases, it could be of use with lit-html as a server-side rendering library

# lang

The `lang` prop on the `<html>` element must be a valid IETF's BCP 47 language tag.

## Rule details

This rule takes no arguments.

### Succeed

```jsx
<html lang="en">
<html lang="en-US">
```

### Fail

```jsx
<html>
<html lang="foo">
```

## Accessibility guidelines
- [WCAG 3.1.1](https://www.w3.org/WAI/WCAG21/Understanding/language-of-page)

### Resources
- [axe-core, valid-lang](https://dequeuniversity.com/rules/axe/3.2/valid-lang)
- [Language tags in HTML and XML](https://www.w3.org/International/articles/language-tags/)
- [IANA Language Subtag Registry](https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry)

**Comment by simonhoyos** [comment]
@bennypowers can I help with this one?

**Comment by bennypowers** [comment]
Please do!

Note that we're focused on release at the moment, so while we may be unable to merge your PR immediately (depending on the timing), we will try to approve it in time, once it passes our review 🙌

**Comment by simonhoyos** [comment]
@bennypowers sounds good. I'll start working on it and will open a PR ASAP

**Comment by simonhoyos** [comment]
@bennypowers Sorry I've been busy these last two weeks and I haven't been able to work as much as I would have wanted on this issue. If that's ok for, I think I can have a PR ready during this week

**Comment by andrico1234** [comment]
Hey @bennypowers I noticed there hasn't been any conversation on this issue for a while, so I thought I'd give it a shot. 

I've started getting the base in place but had a few questions that came up during my research. 

## SSR
You mentioned that this rule will likely provide value for those using lit to render on the server-side. 

I've looked at the lit documentation, and can't find anything about setting the locale/language of the html element. Would we lint instances like this `[Hydrating LitElements](https://www.npmjs.com/package/@lit-labs/ssr)` example? Where, the `<html>` element is contained within the `html` template expression?

Are there any other examples that our linting rule would cover?

## Build time

Since [rollup](https://github.com/open-wc/open-wc/blob/master/docs/docs/building/rollup.md) lets us swap out the lang attribute at build time, do we need to do a check for that too?

This might be a non-issue

## Language codes

Where do we want to store the list of language valid language codes?

- We could have the string saved in the codebase
- We could see if there's a package in which we could import it, save on maintenance
- Create a new package if it doesn't already exist, which can then be reused elsewhere

**Comment by bennypowers** [comment]
You might be able to use this package https://github.com/wooorm/bcp-47

I haven't looked too deeply into the standards though.

WRT use cases I can't think of anything other than ssr. Maybe html embedded in svg? but that's presumably already part of a document

**Comment by andrico1234** [comment]
Thanks @bennypowers, i'll try and whip something up!

**Comment by thepassle** [comment]
Probably a good idea to check out how jsx-a11y does it as well. Seems straight forward: https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/src/rules/lang.js

They use [language-tags](https://www.npmjs.com/package/language-tags)

**Comment by andrico1234** [comment]
Great shout @thepassle !

**Comment by andrico1234** [comment]
I've made a little progress, but hit a roadblock. 

I've done some digging and was wondering if anyone with a bit more experience with the HTML parser could lend their eye. 

In short, when an instance of the `TemplateAnalyzer` tries to traverse an `<HTML>` element, everything behaves as normal until we update the `this._ast` via the `parse5.parseFragment`. (This happens in the `template-analyzer` file.)

It looks like the `parseFragment` method strips out the HTML element, since it no longer holds a reference to the `html` tag within the `children`'s array.

I'm not sure why this is, but it might be because I've misunderstood how Lit treats the `html` element.

I've opened a WIP PR here.

**Comment by thepassle** [comment]
Took some time to debug this today, and asked some help from @43081j (who I cant seem to tag here). Apparently a fragment (passed to `parseFragment`) can't be a HTML document, so it'll get dropped. The alternative to using `parseFragment` is to use `parse`, but that will add `html/head/body` tags to documents/fragments that dont have a `html/head/body` tag, which is a bit unfortunate for most (if not all) lit-html templates. Not really sure what the best way to go about this would be.

**Comment by andrico1234** [comment]
Thanks @thepassle for taking a look at this. I'll give it another shot and see if I can get something working using `parse`.

**Comment by thepassle** [comment]
I think using `parse` might mess up the reporting of error locations though, because it adds a html/head/body tag to fragments

**Comment by andrico1234** [comment]
I wasn't sure how to best demonstrate if it works as expect, so I annotated a test run with some logging showing which `parse` method was used, along with the loc and element tested.

The rest of the test suite works exactly the same as I added a simple `isRootElement` check to the `TemplateAnalyzer`'s `create` method. 

![image](https://user-images.githubusercontent.com/17087167/131908916-3009306d-99a4-4a1d-8155-d64399d8fed4.png)


The [PR is up for review](https://github.com/open-wc/open-wc/pull/2282) and just requires some documentation. Please let me know which other ways I can test the rule

**Comment by andrico1234** [comment]
Also tested this in the lit start kit and works a treat

![image](https://user-images.githubusercontent.com/17087167/132135562-47035437-5a96-4902-9e96-dd2629b682ce.png)

I've also added documentation and the PR is ready for review 😄

**Comment by thepassle** [comment]
> Also tested this in the lit start kit and works a treat

I'm not so much worried about the reporting on the html tag itself, since it's the first thing in the template. What I'm worried about is that parse5 will add a head/body tag if not there, which may (or may not) mess up the error reporting further along in the template. Could you test with the following code snippet and verify that the error reporting is as expected? (preferably post a screenshot)

```js
const y = html`
  <html>
    ${1}
    <img alt="photo"/>

    <img src=""></img>
    ${[1].map(x => html`
      <img alt="photo"/>

      <img src=""></img>
      ${html`
        <div @click=${foo}></div>
        <img alt="photo"/>
        ${html`<img alt="photo"/>`}
        <img src=""/>
        <img alt="photo"/>
      `}
    `)}
    <img alt="photo"/>

    <div @click=${foo}></div>

    <img alt="photo"/>

    <img src=""/>
  </html>
`;
```
